### PR TITLE
perf(gfql): materialize indexed point-query rows directly

### PR DESCRIPTION
**Pull request by lmeyerov**
Scalar node and directed one-hop queries ending in `rows(...)` construct traversal results and evaluate expressions for very small outputs. This change gathers matching rows through resident indexes and projects them directly on pandas, cuDF, and CPU Polars. Joined projections retain one result per matching edge, including repeated matches.

The routes preserve row and column order, dtypes, aliases, empty schemas, graph metadata, validation, and unsupported-query fallback. Polars avoids redundant singleton operations, schema inspection, and validation. Supported equality filters on two to 32 eager Polars rows avoid expression plans; CPU Polars gathers of two to eight rows use slices, with scattered gathers limited to 32 columns. Larger and unsupported cases retain the generic path. Valid property indexes can be reused across CPU and GPU execution targets while retaining frame identity and fingerprint checks.

Review hardening includes:

- Shared identifier/expression parsers, narrower scalar/index/backend types, and removal of join type suppressions.
- Immediate Chain validation plus execution checks for mutation, reuse, custom validators, and error ordering.
- Native and generic path ordering, physical duplicate/null rows and supplied starting wavefronts, and HAS-label collision disambiguation before filtering/deduplication.
- Exact nullable Int64/UInt64 identifiers above 2^53, alias-property restoration through entity identity, and consistent empty-edge schemas/bindings.
- Independent boundary/generalization assertions, with result checks active when routes are disabled and engagement checks scoped to their actual engine/shape. Existing suites replay through all nine individual registered routes plus all-off; CI requires every replay to pass and guards against missing modes.

Current published head: `b5703a35c70b3176e4d815adac04bc1bc9252870`. Real-fixture phase probes on both preceding revisions identified variable one-row projection costs in Polars expression collection (roughly 0.12–0.32 ms). This repair builds resolved Cypher property columns directly from their Series. General Polars projection also gathers direct columns and short-circuits coalesce when matching dtypes and null counts prove the chosen column; mixed-null, mixed-type, and unsupported expression cases retain the generic path. Existing expression lowering, temporal checks, row cardinality, and source isolation remain in force.

The repair adds 64 cases to the existing Polars row-pipeline suite, covering empty/single/larger frames, integer precision, null boundaries, NaN, repeated columns, and categorical/enum/list/struct dtypes. The final focused suite passed 323 tests; broader point/boundary/latency integration passed 697 tests including actual cuDF. Full lint and typing for both changed source files passed. The total review addition is now 641 collected parameterized product cases, not independent bugs or additive execution passes.

Validation on this published revision:

- [Product CI](https://github.com/graphistry/pygraphistry/actions/runs/34728697531): all 85 jobs passed. All nine individual route replays plus all-off report zero divergences. All-off: 12,631 passed, 3,030 skipped, 405 expected failures, two non-strict unexpected passes.
- Full DGX compute plus JSON validation: 17,975 passed, 110 skipped, 102 expected failures, 14 non-strict unexpected passes, and two subtests in 501.74 seconds. Official RAPIDS 26.02, cuDF 26.02.01, Polars 1.35.2; CuPy compute and actual Polars GPU collection preflights passed. Transferred source hash and successful guarded runner/workload completion were verified.
- [pyg-bench candidate](https://github.com/graphistry/pyg-bench/actions/runs/34728758614): 53 mandatory checks passed, no skips.
- [TCK main](https://github.com/graphistry/tck-gfql/actions/runs/34728759781): 4,169 passed/six skipped/689 expected failures in each normal/all-nine-off mode.
- [TCK PR199](https://github.com/graphistry/tck-gfql/actions/runs/34728760728): 4,329 passed/86 skipped/689 expected failures in each mode. [PR199](https://github.com/graphistry/tck-gfql/pull/199) adds 240 regression cases for alias identity, nullable identifiers, path order, node selection, and HAS collisions. All candidate logs verify the exact product SHA.

The final latency sentinel passed unchanged absolute and 1.5× historical-baseline drift gates, including correctness, index engagement, and expected fallback. Eight native Polars shapes measured 0.120–0.281 ms against the 0.7 ms limit. IS5 measured 0.192 ms, seeded-hop properties 0.187 ms, and native coalesce 0.165 ms. The alias-collision fallback returned correct results at 23.525 ms. Sampling: eight warmups and 63 measurements.

Broader measurements use product `b5703a35c70b3176e4d815adac04bc1bc9252870` and frozen harness `557665417a8a5f6f0024b5500e1f9997729d37fe` on DGX Spark. All 12 current Polars/pandas runs and 18 retained matched competitor runs passed source, dataset, sampling, resource, and result validation. Competitor runs come from the earlier H557 campaign; they were not rerun with this product change. Every shared expected and actual result hash matches. Sampling: eight warmups, 31 measurements, three independent runs; timings include execution and full record materialization. Tables show median-of-run-medians in milliseconds. A dash means not measured. These are internal SNB-derived results, not official LDBC results.

SF0.1:

| Query | Polars | Kuzu | Neo4j | Memgraph |
| --- | ---: | ---: | ---: | ---: |
| seed-lookup | 1.071 | 2.404 | 3.212 | 0.638 |
| message-content | 0.171 | 0.648 | 1.838 | 0.373 |
| message-creator | 0.250 | 1.501 | 1.906 | 0.458 |
| message-replies | 9.370 | 24.287 | — | — |
| new-topics | 42.621 | 48.958 | — | — |
| recent-replies | 28.736 | 32.008 | 5.431 | 3.954 |

SF1:

| Query | Polars | Kuzu | Neo4j | Memgraph |
| --- | ---: | ---: | ---: | ---: |
| seed-lookup | 1.462 | 2.084 | 3.420 | 0.509 |
| message-content | 0.173 | 0.688 | 2.102 | 0.432 |
| message-creator | 0.247 | 1.778 | 2.021 | 0.452 |
| new-topics | 169.694 | 210.852 | — | — |

Polars leads Kuzu in all ten eligible comparisons. It leads Neo4j on all six point-query comparisons; Neo4j is faster on SF0.1 recent replies. Polars leads Memgraph on message content and creator at both scales; Memgraph is faster on seed lookup at both scales and SF0.1 recent replies. The ±10% ratio band is a reporting convention, not a statistical significance test; SF0.1 recent replies versus Kuzu is close to the boundary. Relative to the original product73565 run medians, message content improved from 0.417→0.171 ms at SF0.1 and 0.762→0.173 ms at SF1; creator improved from 0.513→0.250 ms and 0.708→0.247 ms. Seed lookup increased about 9% at each scale; other measured larger-query changes range from approximately −1% to +7%.

SF0.1 tag cooccurrence is excluded because its selected parameter returns zero rows. SF1 message replies and recent replies were not measured for Polars. Memgraph used a 4 GiB container/3072 MiB engine cap at SF0.1 and a 16 GiB container/12288 MiB engine cap at SF1, consistent across repetitions. The initial failed SF1 import with the smaller engine cap is retained and excluded.

Separate instrumented GPU/index probes passed all seven SF0.1 and four SF1 result checks on the final product revision. GPU collect calls were observed in five SF0.1 queries and two SF1 queries. Message content and creator used eager CPU paths despite the requested GPU target. Seed lookup, message content, and creator all engaged the property index at both scales. A GPU configuration label does not establish GPU execution for every operation. All instrumented timings are excluded from the comparison tables.

Earlier642be latency runs failed relative drift checks. All failures and alternating-pair/phase diagnostics were retained; phase evidence motivated the current direct-projection repair. Current b570 passed the original gates without threshold relaxation. Historical receipts retain their original revision labels.

The harness has correctness/index/route pins and a candidate latency profile requiring all eight native Polars shapes at or below 0.7 ms plus the expected collision fallback. Candidate checks fail on missing or skipped mandatory tests. [pyg-bench PR260](https://github.com/graphistry/pyg-bench/pull/260) adds automatic enforcement when configured. After product merge, actual-master baselines, measured pandas median×1.5 bounds, release configuration activation, and dependent TCK199 validation/merge remain required. They do not use a candidate result relabeled as master.

Colleague-owned aggregation execution/validation and whole-entity metadata work is tracked in [issue2074](https://github.com/graphistry/pygraphistry/issues/2074). Integration must preserve both sets of correctness expectations. Docs2017 follows the product and benchmark release gates.

An isolated merge of this head with current master `24233cb701e4757d9d6d71c9c0394ce9d4393351` is conflict-free. The combined tree passed 697 focused product tests (one skip; 64 Polars-GPU cases deselected locally) and all 240 TCK199 regressions with pandas, Polars, and cuDF enabled. This is local integration evidence; actual merged-master GPU and baseline validation still follows owner merge.

### CODE REVIEW

File: graphistry/compute/chain_specializations/admission.py [hunk 1]
```diff
@@ -70,8 +70,45 @@ def _indexed_kernel_admits(
     if not (seeded_on_binding or how == "property_index" or n_nodes < n_edges):
         return False
     frac = cost_gate_frac(engine)
-    n_frontier = int(seed_nodes[node].nunique()) if not hasattr(seed_nodes, "get_column") \
-        else int(seed_nodes.get_column(node).n_unique())
+    if not hasattr(seed_nodes, "get_column"):
```
Comment (lmeyerov) [inline graphistry/compute/chain_specializations/admission.py]: audit pr for dynamic typing patterns and switch to static: getattr, hasattr, cast, obj, Any, unqualified collections, str instead of literals, ...

File: graphistry/compute/chain_specializations/point_rows.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+from graphistry.compute.chain_fast_paths import (
+    _index_edge_rows, _index_node_rows, _record_native_seed_lane,
+    _resident_node_id_index, _resident_seed_indexes, _seed_node_rows_from_index,
+    _seeded_scalar_filters, _tag_fast_path_alias_frames, _verify_scalar_filters_on_hit,
+)
+from graphistry.compute.typing import DataFrameT, SeriesT
+from .admission import point_rows_admits
+
+_IDENTIFIER = re.compile(r"[A-Za-z_][A-Za-z_0-9]*\Z")
+_COALESCE_PROPERTIES = re.compile(
```
Comment (lmeyerov) [inline graphistry/compute/chain_specializations/point_rows.py]: DRY externs: check if we have an extern that already do _IDENTIFIER / _COALESCE_PROPERTIES and reuse

File: graphistry/compute/chain_specializations/point_rows.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
+"""Resident-index point queries that produce a row table."""
+import re
```
Comment (lmeyerov) [inline graphistry/compute/chain_specializations/point_rows.py]: do we need / did we add the generalization testing for these new specializations?

File: graphistry/compute/dataframe/join.py [hunk 4]
```diff
@@ -318,13 +318,14 @@ def estimate_inner_join_rows(
...
         import polars as pl
 
-        left_counts = left.group_by(left_on).len().rename({"len": left_n})  # type: ignore[operator]
-        right_counts = right.group_by(right_on).len().rename({"len": right_n})  # type: ignore[operator]
-        value = (
+        from graphistry.compute.gfql.lazy import collect
+
+        left_counts = left.lazy().group_by(left_on).len().rename({"len": left_n})  # type: ignore[operator]
+        right_counts = right.lazy().group_by(right_on).len().rename({"len": right_n})  # type: ignore[operator]
+        value = collect(
```
Comment (lmeyerov) [inline graphistry/compute/dataframe/join.py:326]: avoid type ignores

File: graphistry/compute/chain.py [hunk 5]
```diff
@@ -1042,7 +1050,7 @@ def _chain_impl(
         ops = ops.chain
 
     if validate_schema:
-        Chain(ops).validate(collect_all=False)
+        Chain(ops, validate=False).validate(collect_all=False)
```
Comment (lmeyerov) [inline graphistry/compute/chain.py]: audit -- is this unsafe, and do we repeat this elsewhere?

File: graphistry/tests/compute/gfql/routes/corpus.py [hunk 6]
```diff
@@ -33,6 +33,14 @@ def _entry(name: str, shape: Callable[[KeyMap], List[ASTObject]], tags: Tuple[st
...
+    _entry("point node projection", lambda k: [n({"key": k(1)}, name="a"), rows(source="a"), select(["key", ("value", "a.w")])], ("point-rows", "single-node", "projection")),
+    _entry("point node coalesce", lambda k: [n({"key": k(1)}, name="a"), rows(source="a"), select([("key", "a.key"), ("value", "coalesce(a.w, a.key)")])], ("point-rows", "single-node", "projection")),
+    _entry("point joined hop projection", lambda k: [n({"key": k(1)}, name="a"), e_forward({"type": "KNOWS"}), n(name="b"), rows(), select([("key", "b.key"), ("seed", "a.w"), ("tail", "b.w")])], ("point-rows", "single-hop", "projection")),
+    _entry("point typed hop rows", lambda k: [n({"key": k(1)}, name="a"), e_forward({"type": "KNOWS"}), n(name="b"), rows(source="b")], ("point-rows", "single-hop")),
+    _entry("point typed hop seed rows", lambda k: [n({"key": k(1)}, name="a"), e_forward({"type": "KNOWS"}), n(name="b"), rows(source="a")], ("point-rows", "single-hop")),
+    _entry("point reverse hop projection", lambda k: [n({"key": k(1)}, name="a"), e_reverse({"type": "KNOWS"}), n(name="b"), rows(source="b"), select(["key", ("value", "b.w")])], ("point-rows", "single-hop", "reverse", "projection")),
+
```
Comment (lmeyerov) [inline graphistry/tests/compute/gfql/routes/corpus.py:47]: i'm still a bit weirded out on these -- i thought we already have a lot of tests elsewhere, so why are we making a new corpus.py, instead of more directly broadcasting to specializations? maybe good reason, but if here, feels like we're undercounting majorly?

File: graphistry/compute/chain_specializations/admission.py (hunk 1 above)
Comment (lmeyerov) [reply graphistry/compute/chain_specializations/admission.py]: Tightened the hot-path contracts in 62aa1cd1a: scalar admission returns ScalarFilterDict, the node lookup reason is a Literal, index context/start-node arguments have concrete types, and backend narrowing replaces dynamic attribute probing. Shared dataframe dispatch still needs localized return casts because DataFrameT is currently the pandas checking-time alias. Also repaired five nearby inherited mypy diagnostics. Mypy passes all 15 review-modified source modules. The follow-up 58f3cf1c0 passes the complete repository lint command, including type/comment guards; no caps or suppressions were added. [Scalar contracts](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/compute/typing.py), [backend admission](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/compute/chain_specializations/admission.py).

Final published head `b5703a35c70b3176e4d815adac04bc1bc9252870`: [all 85 product CI jobs passed](https://github.com/graphistry/pygraphistry/actions/runs/34728697531); full actual-DGX compute/JSON validation passed 17,975 tests (110 skips, 102 expected failures, 14 non-strict unexpected passes, two subtests). Actual Polars GPU preflight passed. Unchanged latency gates, all 12 broader GFQL benchmark runs, and both GPU/index diagnostic runs passed. [The PR description](https://github.com/graphistry/pygraphistry/pull/2072) reports exact companion CI, performance, exclusions, and CPU/GPU engagement.

File: graphistry/compute/chain_specializations/point_rows.py (hunk 2 above)
Comment (lmeyerov) [reply graphistry/compute/chain_specializations/point_rows.py]: Replaced both private identifier regexes with is_bare_identifier and the coalesce regex with the shared cached parse_expr grammar. Admission checks the parsed function name, arity, DISTINCT flag, and two bare property operands; unsupported expressions retain fallback. Added 19 lexical boundary cases for case/whitespace, null behavior, nested/quoted expressions, wrong aliases, and malformed calls. Native point tests passed 418 cases with actual cuDF, plus one expected failure. [Implementation](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/compute/chain_specializations/point_rows.py), [tests](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/tests/compute/chain_specializations/test_point_rows.py).

Final published head `b5703a35c70b3176e4d815adac04bc1bc9252870`: [all 85 product CI jobs passed](https://github.com/graphistry/pygraphistry/actions/runs/34728697531); full actual-DGX compute/JSON validation passed 17,975 tests (110 skips, 102 expected failures, 14 non-strict unexpected passes, two subtests). Actual Polars GPU preflight passed. Unchanged latency gates, all 12 broader GFQL benchmark runs, and both GPU/index diagnostic runs passed. [The PR description](https://github.com/graphistry/pygraphistry/pull/2072) reports exact companion CI, performance, exclusions, and CPU/GPU engagement.

File: graphistry/compute/chain_specializations/point_rows.py (hunk 3 above)
Comment (lmeyerov) [reply graphistry/compute/chain_specializations/point_rows.py]: Added 84 independent bag-result cases across sizes 0/1/2/8/9/32/33, direction, aliases, and pandas/Polars/cuDF. They compare admitted execution and all-routes-off execution to expected rows computed from the input edges. Extended existing gather tests with 32 empty/singleton boundary cases and replayed existing chain/collision/reentry suites under the new disabled routes. [Bag matrix](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/tests/compute/gfql/routes/test_point_boundaries.py), [existing-suite replay](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/bin/test-routes-off.sh). Later hardening adds exact order/physical-row/HAS-collision checks and 64 projection cases for empty/single/larger frames, null patterns, mixed-type fallback, repeated columns, and nonprimitive dtypes. Total additions since73565 are 641 collected parameterized product cases, not independent bugs or additive execution passes.

Final published head `b5703a35c70b3176e4d815adac04bc1bc9252870`: [all 85 product CI jobs passed](https://github.com/graphistry/pygraphistry/actions/runs/34728697531); full actual-DGX compute/JSON validation passed 17,975 tests (110 skips, 102 expected failures, 14 non-strict unexpected passes, two subtests). Actual Polars GPU preflight passed. Unchanged latency gates, all 12 broader GFQL benchmark runs, and both GPU/index diagnostic runs passed. [The PR description](https://github.com/graphistry/pygraphistry/pull/2072) reports exact companion CI, performance, exclusions, and CPU/GPU engagement.

File: graphistry/compute/dataframe/join.py (hunk 4 above)
Comment (lmeyerov) [reply graphistry/compute/dataframe/join.py:326]: Removed the six added join-helper type ignores by narrowing Polars frames and checking matching eager/lazy inputs. Eager/lazy semijoin results share one typed union and one dispatcher return cast in 58f3cf1c0. Full repository lint and mypy pass; all 18 existing backend-contract tests pass with actual cuDF enabled. [Join helpers](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/compute/dataframe/join.py), [tests](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/tests/compute/gfql/test_join_backend_contracts.py).

Final published head `b5703a35c70b3176e4d815adac04bc1bc9252870`: [all 85 product CI jobs passed](https://github.com/graphistry/pygraphistry/actions/runs/34728697531); full actual-DGX compute/JSON validation passed 17,975 tests (110 skips, 102 expected failures, 14 non-strict unexpected passes, two subtests). Actual Polars GPU preflight passed. Unchanged latency gates, all 12 broader GFQL benchmark runs, and both GPU/index diagnostic runs passed. [The PR description](https://github.com/graphistry/pygraphistry/pull/2072) reports exact companion CI, performance, exclusions, and CPU/GPU engagement.

File: graphistry/compute/chain.py (hunk 5 above)
Comment (lmeyerov) [reply graphistry/compute/chain.py]: The original expression validates immediately: Chain(..., validate=False).validate(collect_all=False) is equivalent to the validating constructor. Replaced all three occurrences in chain dispatch/native point execution with Chain(ops). The fresh-list validation flag is local; caller-supplied/reused Chains still validate their current mutable operations. Added 56 collected backend cases covering mutation, reuse, deferred validation, custom validators, error ordering, and schema opt-out. [Tests](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/tests/compute/test_chain_validation_execution.py). No validation-bypass defect was established; the full GPU suite now passes as described below.

Final published head `b5703a35c70b3176e4d815adac04bc1bc9252870`: [all 85 product CI jobs passed](https://github.com/graphistry/pygraphistry/actions/runs/34728697531); full actual-DGX compute/JSON validation passed 17,975 tests (110 skips, 102 expected failures, 14 non-strict unexpected passes, two subtests). Actual Polars GPU preflight passed. Unchanged latency gates, all 12 broader GFQL benchmark runs, and both GPU/index diagnostic runs passed. [The PR description](https://github.com/graphistry/pygraphistry/pull/2072) reports exact companion CI, performance, exclusions, and CPU/GPU engagement.

File: graphistry/tests/compute/gfql/routes/corpus.py (hunk 6 above)
Comment (lmeyerov) [reply graphistry/tests/compute/gfql/routes/corpus.py:47]: Agreed that corpus.py is supplemental coverage and cannot establish generalization by itself. Clarified that scope and made bin/test-routes-off.sh replay existing suites using the complete route registry. This also exposed and fixed two gate defects: three newer routes were missing from all-off, and pytest failures could leave a successful shell exit. Two runner regressions now pin future-route inclusion and failure propagation. [Replay runner](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/bin/test-routes-off.sh), [runner tests](https://github.com/graphistry/pygraphistry/blob/b5703a35c70b3176e4d815adac04bc1bc9252870/graphistry/tests/compute/gfql/routes/test_replay.py). [TCK PR198](https://github.com/graphistry/tck-gfql/pull/198) landed the same registry fix. Hosted CI also now includes all nine individual route modes plus all-off, removes job-level failure tolerance, and tests matrix completeness/failure propagation. All ten final-head replays report zero divergences; all-off executes 12,631 passing cases. Result assertions stay active with routes disabled; engagement checks are scoped by engine and route.

Final published head `b5703a35c70b3176e4d815adac04bc1bc9252870`: [all 85 product CI jobs passed](https://github.com/graphistry/pygraphistry/actions/runs/34728697531); full actual-DGX compute/JSON validation passed 17,975 tests (110 skips, 102 expected failures, 14 non-strict unexpected passes, two subtests). Actual Polars GPU preflight passed. Unchanged latency gates, all 12 broader GFQL benchmark runs, and both GPU/index diagnostic runs passed. [The PR description](https://github.com/graphistry/pygraphistry/pull/2072) reports exact companion CI, performance, exclusions, and CPU/GPU engagement.
